ee/clickhouse/queries/person_distinct_id_query.py

```diff
@@ -1,12 +1,27 @@
-from django.conf import settings
+from datetime import timedelta
 
+from ee.clickhouse.materialized_columns.util import cache_for
 from ee.clickhouse.sql.person import GET_TEAM_PERSON_DISTINCT_IDS, GET_TEAM_PERSON_DISTINCT_IDS_NEW_TABLE
+from posthog.models.async_migration import is_async_migration_complete
+from posthog.settings import BENCHMARK, TEST
+
+using_new_table = TEST or BENCHMARK
 
 
 def get_team_distinct_ids_query(team_id: int) -> str:
     from ee.clickhouse.client import substitute_params
 
-    if str(team_id) in settings.PERSON_DISTINCT_ID_OPTIMIZATION_TEAM_IDS:
+    global using_new_table
+
+    using_new_table = using_new_table or _fetch_person_distinct_id2_ready()
+
+    if using_new_table:
         return substitute_params(GET_TEAM_PERSON_DISTINCT_IDS_NEW_TABLE, {"team_id": team_id})
     else:
         return substitute_params(GET_TEAM_PERSON_DISTINCT_IDS, {"team_id": team_id})
+
+
+# :TRICKY: Avoid overly eagerly checking whether the migration is complete.
+@cache_for(timedelta(minutes=15))
+def _fetch_person_distinct_id2_ready() -> bool:
+    return is_async_migration_complete("0003_fill_person_distinct_id2")
```

Comment:
For the write path we don't do this caching, which means we can end up in a situation, where we haven't been writing to pdi original table for 14min, but are still reading from it aka data trust problems. Easier if we double wrote to the old table too, so there's less pitfalls.

Reply:
Fine as is for cloud given how we'll be only landing this later, but problem for self-hosted later if I understood things correctly.

Reply:
Seems like something we can tune. We just want to make sure that we don't over hit this - so maybe we can set this to be `5 minute` cache?

Reply:
I'd vote for just using an env - works for everything. On self hosted they might get perf improvements later (once it's the default) or by adding the env to values, but no data trust problems.

Reply:
Env has the issue where we wait several months for anyone to get any performance wins or requires manual work from their part. This is overcomplicating everything. Will change cache logic around a bit - it incurs extra cost if migration has not run but is mostly free afterwards.